### PR TITLE
Add an aws_hash_table_eq helper

### PR DESCRIPTION
**Pull request by bdonlan**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/hash_table.c [hunk 1]
```diff
@@ -646,6 +646,37 @@ int aws_hash_table_foreach(
     return AWS_OP_SUCCESS;
 }
 
+bool aws_hash_table_eq(
+    bool (*value_eq)(const void *a, const void *b),
+    const struct aws_hash_table *a,
+    const struct aws_hash_table *b) {
+    if (aws_hash_table_get_entry_count(a) != aws_hash_table_get_entry_count(b)) {
```
Comment (ttjsu-aws) [inline source/hash_table.c]: Do we want to add a check for the case when hash_table's are NULL?

Comment (bdonlan) [reply source/hash_table.c]: Generally, we assume that arguments should be non-null except for destructors and for methods that document themselves as accepting null. I'm not sure that null == null is a condition we want to explicitly support here.

File: include/aws/common/hash_table.h [hunk 2]
```diff
@@ -318,6 +318,18 @@ int aws_hash_table_foreach(
...
 
+/**
+ * Compares two hash tables for equality. Both hash tables must have equivalent
+ * key comparators; values will be compared using the comparator passed into this
+ * function. The key hash function does not need to be equivalent between the
+ * two hash tables.
+ */
+AWS_COMMON_API
+bool aws_hash_table_eq(
+    bool (*value_eq)(const void *a, const void *b),
```
Comment (ColdenCullen) [inline include/aws/common/hash_table.h]: Just curious, why is this the first parameter?

Comment (bdonlan) [reply include/aws/common/hash_table.h]: I'm happy with it being either first or last. Which do you think is more consistent?

File: source/hash_table.c [hunk 3]
```diff
@@ -646,6 +646,37 @@ int aws_hash_table_foreach(
...
+        return false;
+    }
+
```
Comment (justinboswell) [inline source/hash_table.c]: If, as the docs say, the key comparators must match, shouldn't that be verified here by comparing equals_fn on both?

File: include/aws/common/hash_table.h (hunk 2 above)
Comment (ColdenCullen) [reply include/aws/common/hash_table.h]: My vote is last, looking at things like C++'s [`std::equal`](https://en.cppreference.com/w/cpp/algorithm/equal), the predicate is always last.

File: source/hash_table.c (hunk 3 above)
Comment (bdonlan) [reply source/hash_table.c]: We don't require that they are _equal_, just that they are _equivalent_. For example, we might pass in two FFI-generated bindings to the same lambda function in an interpreted language.

File: source/hash_table.c (hunk 1 above)
Comment (ttjsu-aws) [reply source/hash_table.c]: I see, thanks for stating the assumption.

**Pull request merged**